src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,276 @@
+/// in-memory buffer, then write that buffer's length to the pipe-writer, then copy the in-memory buffer to the
+/// writer.
+/// </para>
+/// When writing/reading the data-segment, we use an the <see cref="ObjectWriter"/>/<see cref="ObjectReader"/>
+/// subsystem.  This will write its own validation bits, and then the data describing the asset.  This data is:
+/// <code>
+/// -----------------------------------------------------------------------------------------------------------
+/// | data (variable length)                                                                                  |
+/// -----------------------------------------------------------------------------------------------------------
+/// | ObjectWriter validation (2 bytes) | checksum (16 bytes) | kind (4 bytes) | asset-data (asset specified) |
+/// -----------------------------------------------------------------------------------------------------------
+/// </code>
+/// The validation bytes are followed by the checksum.  The checksum is needed in the message as assets can be found
+/// in any order (they are not reported in the order of the array of checksums passed into the writing method).
+/// Following this is the kind of the asset.  This kind is used by the reading code to know which
+/// asset-deserialization routine to invoke. Finally, the asset data itself is written out.
+/// </summary>
+internal static class RemoteHostAssetSerialization
 {
-    internal static class RemoteHostAssetSerialization
+    /// <summary>
+    /// A sentinel byte we place between messages.  Ensures we can detect when something has gone wrong as soon as possible.
+    /// </summary>
+    private const byte MessageSentinelByte = 0b01010101;
+
+    private static readonly ObjectPool<SerializableBytes.ReadWriteStream> s_streamPool = new(SerializableBytes.CreateWritableStream);
+
+    public static async ValueTask WriteDataAsync(
+        PipeWriter pipeWriter,
+        AssetPath assetPath,
+        ReadOnlyMemory<Checksum> checksums,
+        SolutionAssetStorage.Scope scope,
+        ISerializerService serializer,
+        CancellationToken cancellationToken)
     {
-        public static async ValueTask WriteDataAsync(
-            Stream stream,
-            Dictionary<Checksum, object> assetMap,
-            ISerializerService serializer,
-            SolutionReplicationContext context,
-            Checksum solutionChecksum,
-            ReadOnlyMemory<Checksum> checksums,
-            CancellationToken cancellationToken)
+        // Create a channel to communicate between the searching and writing tasks.  This allows the searching task to
+        // find items, add them to the channel synchronously, and immediately continue searching for more items.
+        // Concurrently, the writing task can read from the channel and write the items to the pipe-writer.
+        var channel = Channel.CreateUnbounded<(Checksum checksum, object asset)>(new UnboundedChannelOptions()
```

Comment:
>new UnboundedChannelOptions() [](http://example.com/codeflow?start=81&length=29)

Can this be static and shared across calls?

Reply:
likely so!